lib/heaptrack/heaptrack-wrapper.ts

```diff
@@ -0,0 +1,116 @@
+import * as path from 'path';
+import {mkfifoSync} from 'mkfifo';
+import {ExecutionOptions} from '../../types/compilation/compilation.interfaces.js';
+import {TypicalExecutionFunc, UnprocessedExecResult} from '../../types/execution/execution.interfaces.js';
+import {O_NONBLOCK, O_RDWR} from 'constants';
+import * as fs from 'fs';
+import * as net from 'net';
+import {pipeline} from 'stream';
+import {unwrap} from '../assert.js';
+import {logger} from '../logger.js';
+
+export class HeaptrackWrapper {
+    private dirPath: string;
+    private raw_output: string;
+    private pipe: string;
+    private heaptrackPath: string;
+    private preload: string;
+    private interpreter: string;
+    private printer: string;
+    private sandboxFunc: TypicalExecutionFunc;
+    private execFunc: TypicalExecutionFunc;
+
+    public static FlamegraphFilename = 'heaptrack.flamegraph.txt';
+
+    constructor(dirPath: string, sandboxFunc: TypicalExecutionFunc, execFunc: TypicalExecutionFunc) {
+        this.dirPath = dirPath;
+        this.pipe = path.join(this.dirPath, 'heaptrack_fifo');
+        this.raw_output = path.join(this.dirPath, 'heaptrack.raw');
+        this.heaptrackPath = '/opt/compiler-explorer/heaptrack/1.3.0'; // todo: needs to be a property
+        this.preload = path.join(this.heaptrackPath, 'lib/libheaptrack_preload.so');
+        this.interpreter = path.join(this.heaptrackPath, 'libexec/heaptrack_interpret');
+        this.printer = path.join(this.heaptrackPath, 'bin/heaptrack_print');
+        this.sandboxFunc = sandboxFunc;
+        this.execFunc = execFunc;
+    }
+
+    private async make_pipe() {
+        mkfifoSync(this.pipe, 0o666);
+    }
+
+    private add_to_env(execOptions: ExecutionOptions) {
+        if (!execOptions.env) execOptions.env = {};
+
+        if (execOptions.env.LD_PRELOAD) {
+            execOptions.env.LD_PRELOAD = this.preload + ':' + execOptions.env.LD_PRELOAD;
+        } else {
+            execOptions.env.LD_PRELOAD = this.preload;
+        }
+
+        execOptions.env.DUMP_HEAPTRACK_OUTPUT = '/app/heaptrack_fifo';  // todo: will not work for local users without nsjail, is that a problem?
```

Comment:
https://github.com/compiler-explorer/compiler-explorer/pull/5533/files#diff-23200c2e7936d6738c5056e91063698d8d9db45b30fa99be68e22eed138ee8eaR219 offers code to identify the sandboxtype, should become its own PR that we can merge in here and use for this